utils/env_preferences/src/apple.rs

```diff
@@ -0,0 +1,136 @@
+        base::{CFRelease, CFRetain},
+        calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+        locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+        string::CFStringGetCStringPtr,
+    };
+    use std::ffi::CStr;
+
+    pub fn get_locales() -> Vec<String> {
+        let mut languages: Vec<String> = Vec::new();
+
+        // SAFETY: The call to `CFLocaleCopyPreferredLanguages` returns an immutable reference to `CFArray` which is owned by us
+        // https://developer.apple.com/documentation/corefoundation/cfarrayref. It is ensured that `locale_carr_ref` is not mutated
+        // Immutablility ensures that nothing is overriden during it's scope
+        let locale_carr_ref = unsafe { CFLocaleCopyPreferredLanguages() };
+
+        if !locale_carr_ref.is_null() {
+            // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
+            let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
+            for i in 0..count {
+                // SAFETY: The call to `CFArrayGetValueAtIndex` is safe because we are iterating from 0 to count (`CFArrayGetCount`) which
+                // in itself will always be greater than 0 and less than count ensuring we will not get "out of bounds" error
+                let lang_ptr = unsafe { CFArrayGetValueAtIndex(locale_carr_ref as _, i) };
+
+                if !lang_ptr.is_null() {
+                    // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
+                    // Returns pointer in O(1) without any memory allocation. This can return NULL so to handle it
+                    let lang_str = unsafe { CFStringGetCStringPtr(lang_ptr as _, 0) };
+
+                    if !lang_str.is_null() {
+                        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+                        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }
+                            .to_str()
+                            .unwrap_or("Unknown")
+                            .to_string();
+                        languages.push(lang_rust_str);
+                    }
+
+                    // TODO: In case `lang_str` is `NULL` try retrieving the string using `CFStringGetCString`
+                    // Ref: https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring?language=objc
+                    // Note: Not optimal and may give inconsistent results if buffer is not large enough, must add sanity checks
+                    // whenever implemented
+                }
+            }
+        }
+        // Release for memory
+        unsafe {
+            CFRelease(locale_carr_ref as _);
+        }
+
+        // Defaulting to `und`
+        if languages.is_empty() {
+            languages.push(String::from("und"));
+        }
+        languages
```

Comment:
I would not do this in this low level function. In my mental model:
* Low Level - get information from env. No logic. If languages is empty, return empty. If it fails, return failure.
* High level - normalize results from different platforms into a single cohesive result. Here comes heuristics such as this.